File: 5주차/BOJ_1717_집합의표현/dongjun.java
/**
 * 
 *  @문제접근
 *  UNION-FIND로 풀었습니다.
 *
 *  @성능
 *  Runtime: 552 ms
 *  Memory Usage: 65012 KB
 *  시간복잡도: O(M)
 *  
 */
public class BOJ_1717_집합의표현 {
    static final int UNION = 0, FIND = 1;
    
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

	public static void main(String[] args) throws IOException {
        String[] input = br.readLine().split(" ");

        solve(Integer.parseInt(input[0]),Integer.parseInt(input[1]));
    }

    private static void solve(int n, int m) throws IOException {
        int[] set = new int[n+1];
        Arrays.fill(set, -1);
        int order, a, b;

        for(int i = 0; i < m; i++){
            String[] input = br.readLine().split(" ");
    
			order = Integer.parseInt(input[0]);
            a = Integer.parseInt(input[1]);
            b = Integer.parseInt(input[2]);

            if(order == UNION)
                union(a, b, set);
            else {
                if(isSameSet(a, b, set)){
                    bw.write("YES\n");
                }
                else
                    bw.write("NO\n");
            }
        }
        bw.flush();
        br.close();
        bw.close();
    }

    private static void union(int a, int b, int[] set) {
        int parentA = find(a, set);
        int parentB = find(b, set);
        if(parentA == parentB)
            return;
        set[parentB] = parentA;
    }

    private static int find(int a, int[] set) {
        if(set[a] < 0)
            return a;
        return set[a] = find(set[a], set);
    }

    private static boolean isSameSet(int a, int b, int[] set) {
        return find(a, set) == find(b, set);
    }
}

File: 9주차/BOJ_9465_스티커/dongjun.py
"""
1. 문제접근
    동적계획법으로 풀었습니다.
    현재 스티커를 포함하는 경우와 포함하지 않는 경우로 나눠 둘 중에 큰 값을 저장하여 풀었습니다.

2. 성능
    Runtime: 796ms
    Memory Usage: 39768KB
"""
def main():
    T = int(input())

    for tc in range(1, T+1):
        n = int(input())
        stickers = [list(map(int, input().split())) for _ in range(2)]
        print(solve(n, stickers))


def solve(n, stickers):
    TOP, BOTTOM = 0, 1
    dp = [[0]*n for _ in range(2)]
    dp[TOP][0] = stickers[TOP][0]
    dp[BOTTOM][0] = stickers[BOTTOM][0]

    for i in range(1, n):
        dp[TOP][i] = max(stickers[TOP][i] + dp[BOTTOM][i-1], dp[TOP][i-1])
        dp[BOTTOM][i] = max(stickers[BOTTOM][i] + dp[TOP][i-1], dp[BOTTOM][i-1])

    return max(dp[TOP][n-1], dp[BOTTOM][n-1])


if __name__ == '__main__':
    main()
